src/System.Management.Automation/engine/NativeCommandProcessor.cs

```diff
@@ -1344,676 +1431,434 @@ private bool IsMiniShell()
+            _streamWriter = new StreamWriter(process.StandardInput.BaseStream, pipeEncoding);
+            _streamWriter.AutoFlush = true;
 
-        /// <summary>
-        /// Launches a new thread to start reading.
-        /// </summary>
-        internal void Start()
-        {
-            _thread = new Thread(new ThreadStart(ReaderStartProc));
-            if (_isOutput)
+            _inputFormat = inputFormat;
+
+            if (_inputFormat == NativeCommandIOFormat.Xml)
             {
-                _thread.Name = string.Format(CultureInfo.InvariantCulture, "{0} :Output Reader", _processPath);
+                _streamWriter.WriteLine(ProcessOutputHandler.XmlCliTag);
+                _xmlSerializer = new Serializer(XmlWriter.Create(_streamWriter));
             }
-            else
+            else // Text
             {
-                _thread.Name = string.Format(CultureInfo.InvariantCulture, "{0} :Error Reader", _processPath);
+                _pipeline = ScriptBlock.Create("Out-String -Stream").GetSteppablePipeline();
+                _pipeline.Begin(true);
             }
-            _thread.Start();
         }
 
+        bool _stopping = false;
+
         /// <summary>
-        /// This method returns when reader thread has returned.
+        /// Stop writing input to process
         /// </summary>
-        internal void Done()
+        internal void Stop()
         {
-            if (_thread != null)
-            {
-                _thread.Join();
-            }
+            _stopping = true;
         }
 
         /// <summary>
-        /// Thread proc for reading
+        /// This method wait for writer thread to finish.
         /// </summary>
-        private void ReaderStartProc()
+        internal void Done()
         {
-            try
-            {
-                ReaderStartProcHelper();
-            }
-            catch (Exception ex)
-            {
-                CommandProcessorBase.CheckForSevereException(ex);
-            }
-            finally
+            // we allow call Done() multiply times.
+            // For example one time from ProcessRecord() code path,
+            // when we detect that process already finished
+            // and once from Complete() code path.
+            // Even though Done() could be called multiple times,
+            // the calls are on the same thread, so there is no race condition.
+            if (_pipeline != null)
             {
-                _processOutputReader.ReaderDone(_isOutput);
+                var finalResults = _pipeline.End();
+                _pipeline.Dispose();
+                _pipeline = null;
+                // AddTextInputFromFormattedArray can recursively call Done(),
+                // if the downstream process already exited.
+                // to Prevent it, we first finalize the pipeline and set it to null,
+                // then calling the result processing for the last time
+                AddTextInputFromFormattedArray(finalResults);
```

Comment:
speaking of try/catch perf penalty, when `Done()` is called from `AddTextInputFromFormattedArray`, this will raise another exception for sure, which should be avoided.
Two questions:
1. Is calling `_pipeline.End()` really absolutely necessary? It seems `_pipeline.End()` for `Out-String -stream` only write out some newlines. Is there a regression without those new lines?
2. If it's really necessary, I suggest having `Done` renamed to be `Dispose`:
```
internal Dispose()
{
    if (_pipeline != null)
    {
        _pipeline.Dispose();
        _pipeline = null;
    }

    // ... Same as before in below.
}
```
Then have another method named `Done()`
```
internal Done()
{
    if (_pipeline != null)
    {
        AddTextInputFromFormattedArray(_pipeline.End());
    }
    Dispose()
}
```
So in `AddTextInput` and `AddXmlInput`, it calls `Dispose()` when the IO exception is caught, while in `Complete`, it calls `Done()`.

Another alternative is to add a parameter to the current `Done()` method, so that if it's called from `AddTextInput` or `AddXmlInput`, don't call `_pipeline.End()` and `AddTextInputFromFormattedArray`